### ISSUE TITLE
ENH: support bluesky callbacks

### ISSUE THREAD

**Issue by CJ-Wright** [xpdAcq/redsky#6]
It would be good to support bluesky callbacks, allowing data to be processed as it comes from the detector.  The main issue here may be that we need to open a bunch of run starts without run stops.

Here is one potential workflow:
1. Create a master callback which contains all the internal `RunFunction` instances (the task graph, essentially) 
1. On `start` issue all the run starts and descriptors
   1. We will need to trigger each start in sequence and pass it to the next one as needed.
1. On `event` run the chain of `AnalysisRunEngine` commands
1. On `stop` issue all the run stops
This approach will require the separation of the various steps so that we can subclass from bluesky callbacks.  Essentially each of the `AnalysisRunEngine` calls is held in suspense waiting for the next piece of data to roll in. A massive upside of this is that one could, potentially, use `db.process(master_callback)` to perform the entire workflow, in addition to having it done on the spot.

Potential issues with this approach:
* This may be not very error tolerant if the analysis generator breaks then the pipeline may stop. (There seems to be some setup for error toleration inside the callbacks, although I don't know what happens inside the task graph (error cascade?))
* I am not certain how to handle the event data, currently we always operate on complete data. Now we will need to operate on each individual event separately. This is actually a boon in disguise, since we can do this without problem in the callback setup and the dask calls will be auto parallel (and even cached if we play our cards right).
* Analysis which requires multiple headers might be difficult. Need to think about how to do dark association. This is not so bad so long as we are inside a dask computation. Outside is worse since we need to multiplex the data together.
* All of this needs to happen somewhere else, since we can't afford to slow down the control computer (this may be mostly solved by [zmq based callbacks](http://nsls-ii.github.io/bluesky/callbacks.html#subscriptions-in-separate-processes-or-host-with-0mq))
* `db.process(sub_master_callback)` may have some problems with processing steps which require multiple headers (this may be where a special reprocess function may need to be, this would look at the parent headers and properly construct a stream(s) to be friendly to the dask task graph)
* I am not certain how to inject visualization/gui code in such a way that the first event to be processed runs the gui, a value/set of values for data processing are picked, and the rest of the events are processed with those values.

**Comment by CJ-Wright** [comment]
This should be combined with #2.
for example:
```python
from bluesky.callbacks.core import CallbackBase

class MasterDaskCallback(CallbackBase):
    def __init__(self, dask_task_graph):
        # prep dask task graph, may need to leave room for 'name' in all the calls
        # throw task graph serialization into database somewhere???
        dsk1 = dask_task_graph
    def start(self, doc):
        # run the task graph with the name 'start' in the name portion
        dsk1.compute()
    def descriptor, event, stop(self, doc):
        # same as above

class ProcessAndStore(CallbackBase):
    def __init__(self, stuff):
        # put in an analysis databroker here
    def start(self, doc):
        # Same for the others, issue the correct document based off of the data coming down
        # special attention paid for the event data when everything is processed

**Comment by CJ-Wright** [comment]
@tacaswell @sbillinge You might take a look at this if you have the time, it is a potential alternative to the existing PRs which may be a bit more flexible. It is still very much a work in progress and has arisen from my attempts to work analysis database saving with bluesky callbacks. I think it may address some of the issues you were raising (eg using the streams rather than the direct headers and being less dependent on knowing the databroker from where all our data comes from and native dask integration, although some things are still issues; we still need to know some extra info about our internal functions).

**Comment by CJ-Wright** [comment]
While I've been working on a generator/stream based data processing I've been thinking about Callback based data processing. The Callback based method seems to have the advantage that we could use the same codebase for both live and post-acquisition processing, where generator/stream based data processing might not work with live data (since the live data is not in a generator).

The idea would be to break the callback framework up into 4 parts:
1. `db_store` decorator: this handles the saving of documents into MDS and FS
2. math functions: these perform the base level transformations
3. `AnalysisCallbacks`: These map the documents to the math functions and return new documents with the results
4. `ExecutorCallbacks`: These provide the order and flow to the various `AnalysisCallbacks` passing name, document pairs between the `AnalysisCallbacks`. This provides callback chaining

There is also a class decorator which decorates the `__call__` method from `CallbackBase` with a `db_store` instance which makes certain that everything is properly stored

Here are some nasty corner cases that seem to pop up:
1. `CollectAndCompute`: This causes a problem because the `event` document can only be issued at the end when the `stop` document is received. Thus the `stop` document is never issued.
2. `FiterCallbacks` which don't issue on every doc or issue more docs than the `RunEngine`: Consider a callback which filters events. This callback will not always issue new event docs (an event has been filtered out) but what happens to callbacks waiting downstream? Similarly, what happens if that for every one doc that goes in two come out. What happens to the not created docs when the scan finishes?
   1. I guess we could write things so that filters return `(None, None)` when they don't want to issue an event and have all Callbacks understand that they should just pass  `(None, None)` down the chain and do nothing. This doesn't solve the issue with issuing many events per 1 input event.

@tacaswell @danielballan as always I am very interested in your thoughts.

**Comment by CJ-Wright** [comment]
Closed by #13
